src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -87,13 +225,91 @@ public override TypedExpression Transform(TypedExpression ex)
+            if (types != null && !types.IsEmpty)
+            {
+                target = CreateConcretion(types);
+            }
+
+            string name = null;
+
+            // Check for recursive call
+            if (CurrentResponse.originalName.Namespace.Value == globalCallable.Item.Namespace.Value &&
+                CurrentResponse.originalName.Name.Value == globalCallable.Item.Name.Value &&
+                CurrentResponse.typeResolutions.SetEquals(target))
+            {
+                name = CurrentResponse.concreteCallable.FullName.Name.Value;
+            }
+
+            // Search Requests for identifier
+            if (name == null)
+            {
+                name = Requests
+                    .Where(req =>
+                        req.originalName.Namespace.Value == globalCallable.Item.Namespace.Value &&
+                        req.originalName.Name.Value == globalCallable.Item.Name.Value &&
+                        req.typeResolutions.SetEquals(target))
+                    .Select(req => req.concreteName.Name.Value)
+                    .FirstOrDefault();
+            }
+
+            // Search Responses for identifier
+            if (name == null)
+            {
+                name = Responses
+                    .Where(res =>
+                        res.originalName.Namespace.Value == globalCallable.Item.Namespace.Value &&
+                        res.originalName.Name.Value == globalCallable.Item.Name.Value &&
+                        res.typeResolutions.SetEquals(target))
+                    .Select(res => res.concreteCallable.FullName.Name.Value)
+                    .FirstOrDefault();
+            }
+
+            // If identifier can't be found, make a new request
+            if (name == null)
+            {
+                if (target != null)
+                {
+                    // Create new name
+                    name = Guid.NewGuid().ToString() + "_" + globalCallable.Item.Name.Value;
+
+                    // TODO: Check for identifier name collisions (not likely)
+                    // - check in global namespace
+                    // - check in concretion dict for current generic
+                }
+                else
+                {
+                    // If this is not a generic, do not change the name
+                    name = globalCallable.Item.Name.Value;
+                }
+
+                Requests.Push(new Request()
+                {
+                    originalName = globalCallable.Item,
+                    typeResolutions = target,
+                    concreteName = new QsQualifiedName(globalCallable.Item.Namespace, NonNullable<string>.New(name))
+                });
+            }
+
+            return Identifier.NewGlobalCallable(new QsQualifiedName(globalCallable.Item.Namespace, NonNullable<string>.New(name)));
         }
     }
+
+    internal class ReplaceTypeParamCallsExpressionKind : ExpressionKindTransformation<ReplaceTypeParamCallsExpression>
```

Comment:
I am assuming you will add functionality to this class later on?

Reply:
Yes this class is filled out in PR #206 . Strangely though, when I didn't have this here, even empty as it is, the transform wouldn't recurse properly, and having this here fixed the issue.